pkg/controller/seed-controller-manager/ipam/helper.go

```diff
@@ -0,0 +1,130 @@
+/*
+Copyright 2022 The Kubermatic Kubernetes Platform contributors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ipam
+
+import (
+	"fmt"
+	"math/big"
+	"net"
+)
+
+var (
+	errIncompatiblePool = fmt.Errorf("pool is incompatible with a current cluster allocation")
+)
+
+type datacenterIPAMPoolUsageMap map[string]struct{}
+
+func newDatacenterIPAMPoolUsageMap() datacenterIPAMPoolUsageMap {
+	return make(datacenterIPAMPoolUsageMap)
+}
+
+func (m datacenterIPAMPoolUsageMap) setUsed(value string) {
+	m[value] = struct{}{}
+}
+
+func (m datacenterIPAMPoolUsageMap) isUsed(value string) bool {
+	_, isUsed := m[value]
+	return isUsed
+}
+
+func ipToInt(ip net.IP) (*big.Int, int) {
+	val := &big.Int{}
+	val.SetBytes([]byte(ip))
+	switch len(ip) {
+	case net.IPv4len:
+		return val, 32
+	case net.IPv6len:
+		return val, 128
+	default:
+		panic(fmt.Errorf("unsupported address length %d", len(ip)))
+	}
+}
+
+func intToIP(ipInt *big.Int, bits int) net.IP {
+	ipBytes := ipInt.Bytes()
+	ret := make([]byte, bits/8)
+	// Pack our IP bytes into the end of the return array,
+	// since big.Int.Bytes() removes front zero padding.
+	for i := 1; i <= len(ipBytes); i++ {
+		ret[len(ret)-i] = ipBytes[len(ipBytes)-i]
+	}
+	return net.IP(ret)
+}
+
+func addressRange(network *net.IPNet) (net.IP, net.IP) {
+	// the first IP is easy
+	firstIP := network.IP
+
+	// the last IP is the network address OR NOT the mask address
+	prefixLen, bits := network.Mask.Size()
+	if prefixLen == bits {
+		// Easy!
+		// But make sure that our two slices are distinct, since they
+		// would be in all other cases.
+		lastIP := make([]byte, len(firstIP))
+		copy(lastIP, firstIP)
+		return firstIP, lastIP
+	}
+
+	firstIPInt, bits := ipToInt(firstIP)
+	hostLen := uint(bits) - uint(prefixLen)
+	lastIPInt := big.NewInt(1)
+	lastIPInt.Lsh(lastIPInt, hostLen)
+	lastIPInt.Sub(lastIPInt, big.NewInt(1))
+	lastIPInt.Or(lastIPInt, firstIPInt)
+
+	return firstIP, intToIP(lastIPInt, bits)
+}
+
```

Comment:
I sure would love some basic unit tests for these helpers 😁

Reply:
It's a bit confusing to have the unit tests be part of another, future PR. But okay...

Reply:
I added some tests already: [Adding helper tests](https://github.com/kubermatic/kubermatic/pull/10028/commits/bebe0adb87e43c0b371aabf4c6ae81dca2a3fafd)